Add multiplayer support and fix edge cases

JamieNyanchi opened the pull request:
This took me longer than I expected, but here's my code as promised! I tried to give a brief explanation below of what I did. I have also tried to test all these changes as well as I could (which is why this PR took so long). If you'd like more information or for me to make any changes, just let me know! Thank you!

---
<details>
<summary>1. Add multiplayer support (Outdated, see comments)</summary><br>
This is the where the majority of the changes in the PR are from. As noted on the workshop by many other players the map tacks only work properly for the host in multiplayer. This is because the map tack data is saved in the player properties, but this data is overwritten by the host's copy of the properties very often. Since each client calculates the data and saves it themselves, it would be lost once the host sent their data. <br><br>
To fix this, I primarily took advantage of `UI.RequestPlayerOperation`. This is a UI function that can call a script/gameplay function, much like ExposedMembers can. However, this function has the advantage of calling that function for the player on every client. I use this to essentially send the map tack data to every other client. This also means that when players make their map tacks visible to others, the map tack details show for the others too.<br><br>
There are some edges cases here, like if a player places pins down while their turn isn't active, but I accounted for those. In cases like those, it isn't possible for this function to send the data to the host until the player's turn is active again. If that happens, the mod will "cache" the current state of their map tacks, reapply them when their turn is active again, and then send the data to the host. If the game is saved and exited in this state, then the map tack data will be fully recalculated for that player when the next session is started. I tested this with a friend quite a bit and from our tests, it seems the map tack data is properly preserved across turns and across sessions. <br><br>
Resolves #2 
</details>

---
<details>
<summary>2. Fix map tack details when placed on edges of map</summary><br>
This was done just by replacing `ipairs` with `pairs` when iterating over the `Map.GetAdjacentPlots` function return. This function can return an array with nil plots, and `ipairs` only iterates until it hits a nil entry. As a result, when iterating over the return value with `ipairs`, not all the tiles would be checked if one happens to be nil, and incorrect values could be displayed on the map tack. <br><br>
As a quick example, placing a campus at the very top of the map with one mountain adjacent to it should should show +1 as its adjacency, but it instead shows as nothing (or +0). After this change, it shows the correct values. <br><br>
From my understanding, this should not increase the chances of desyncing in multiplayer. In fact, with my other changes, there should be a reduction in the chances for desyncing. If I understand correctly, using `pairs` instead of `ipairs` is only an issue in script/gameplay files, not UI files. After all my other changes, there are no more script/gameplay files using `pairs`, so this shouldn't be an issue.
</details>

---
<details>
<summary>3. Add compatibility with CQUI</summary><br>
This small addition is optional to be honest. CQUI doesn't do anything that Detailed Map Tacks doesn't already do. This just allows CQUI to add in their own changes to the map tacks feature themselves, if they wanted to. I have ensured that compatibility with the other Map Tacks mod is still maintained. No compatibility is lost here, and the mod works fine with or without the other map tacks mod and/or CQUI.
</details>

---

hanweizhang commented on ui/dmt_yieldcalculator.lua:
```diff
@@ -81,19 +87,41 @@ function UpdatePinYields(playerID:number, pinsToUpdate:table)
+        -- Update the config if the state should be changed
+        if (not bSuccess and restoreSubjectsConfig ~= 1) then
+            playerConfig:SetValue(RESTORE_MAP_PIN_SUBJECTS_CONFIG_KEY, 1);
+            Network.BroadcastPlayerInfo();
```
If playerConfig updates can be broadcasted to all clients through `Network.BroadcastPlayerInfo()`, can we simply store player's all MapPinSubjects to playerConfig? I'm not familiar with multi player mode at all, but looking through the way you are using it, it doesn't seem like it will ever be overridden by host's copy. I haven't worked on Civ6 modding for a while, I don't remember if playConfig can store complex values like tables in this case. If not, we can use one of serialize/deserialize options here http://lua-users.org/wiki/TableSerialization.

Also I suppose `Network.BroadcastPlayerInfo()` is more reliable than `UI.RequestPlayerOperation` such that we don't need the local cache/restore logic?

JamieNyanchi replied:
I'm glad you brought that up, because I neglected to mention that in my original post. When I originally was researching how to go about this, I thought about doing that initially. Like you suspected though, you can't store tables in the player configuration. I decided to go with the solution I originally had here because I was worried about the potential performance impact of the serialization / deserialization functions.

After reading your comment though, I reconsidered the idea and I agree with you. The solution I originally had here became more complex than I originally thought it would, and ended up having some issues come up later on. It mostly worked, but had some issues occurring seemingly randomly (probably due to how Civ 6 handles networking?). These issues go away when saving to the player configuration instead though. In addition, the player configuration solution is much simpler since it doesn't need the cache/restore logic, like you pointed out.

So, after reconsidering, I replaced my original solution here with the player configuration idea. I haven't tested it thoroughly yet, so I'll leave it as a draft until I do in the next couple days. I have tested the performance a little bit though. The serialization / deserialization functions I used seem to make no noticeable impact on performance.

The `Network.BroadcastPlayerInfo()` function seems to have a very minor impact on performance in multiplayer games, but only by a small fraction of a second at most when placing/editing pins. The impact it has varies and sometimes is negligible. To keep it as minimal as possible though, I only call it at the end of the `UpdatePinYields()` function. Due to how the base game MapPinManager is written, calling this function also refreshes all the pins on the map when it is called. As such, I was able to remove the `RefreshMapPinUI` LuaEvent, and everything still updates as expected.

Sorry for the wall of text, but I wanted to explain all my thoughts there. I look forward to hearing your thoughts if you have any!

hanweizhang commented on ui/dmt_yieldcalculator.lua:
```diff
@@ -494,7 +522,7 @@ end
     -- Cannot use plot:IsAdjacentToLand() since it'll include unrevealed plots.
     local adjPlots = Map.GetAdjacentPlots(x, y);
-    for i, plot in ipairs(adjPlots) do
+    for i, plot in pairs(adjPlots) do
```
Ack `ipairs` changes to `pairs` to support top/bottom of the map. May need to double check for NPEs after the change.

If order is needed in any of these use cases, maybe we can also define a new `ipairsNilSafe` method by removing the if check in `iter` here https://www.lua.org/pil/7.3.html

JamieNyanchi replied:
I looked back through where I changed `ipairs` to `pairs` and it looks like it's all good in terms of NPEs. All of them check if the `plot` value is nil or not, so that part should be covered.

If I understand the existing code correctly, it doesn't seem like the order matters at all in any of these cases. However, if I'm wrong and it does matter somewhere, we could certainly do an alternative. We could try the one you proposed, or maybe do our own `GetAdjacentPlots` function. From my experience, it seems the `Map.GetAdjacentPlots()` function returns them in counter-clockwise order starting from the plot to the top left of the center plot. Replicating this behavior should be trivial.

If any case, I haven't noticed any issues in regards to these particular changes in my playtests, but I'd be happy to make some adjustments if you'd like!

hanweizhang commented on ui/mappinmanager_dmt.lua:
```diff
@@ -179,19 +188,26 @@ function OnAutoDeleteRememberChoice(checked)
 function OnInputHandler(pInputStruct:table)
+    if (BASE_OnInputHandler) then
+        BASE_OnInputHandler(pInputStruct);
+    end
```
Ack the compatibility change.

hanweizhang reviewed (commented):
Appreciate your contribution. I learned a lot from your comments too, thanks!

JamieNyanchi committed:
Add multiplayer support and fix edge cases

1. Add multiplayer support

2. Fix map tack details when placed on edges of map

3. Add compatibility with CQUI

JamieNyanchi commented:
Sorry for the delay, but I'm finally comfortable with the new implementation here to mark this PR as ready to review. I wanted to be careful and do a lot of tests to ensure there are no regressions or anything, especially since this is literally the most subscribed mod on the Civ 6 workshop.

I described the implementation a bit in responses to your comments above, but here's a quick recap of how this works now. Map Pin data is now saved to the player configuration as a serialized table (since the player configuration doesn't allow tables directly). The UI is now updated through use of the `Network.BroadcastPlayerInfo()` command, as the base game is set up to automatically refresh every map pin when the player config is updated.

In addition, I have added in a cache system to avoid performance regressions with the new system. When the map pins are to be updated, a cache is created is hold all of the changes until everything is finished updating. At that point, it is saved to the player config and broadcasted to the other clients. This cache system isn't needed at all for accuracy or anything, and is purely for performance. Without it (and only in multiplayer games), adding/removing/updating a pin could take up to half a second for it to update on the UI. With the cache, it appears to be just as fast as it was before. Singleplayer never appeared to have a performance impact, so that's unaffected by these changes.

Hopefully you agree that this is a better implementation! I look forward to your thoughts and comments!

hanweizhang commented on ui/dmt_mappinsubjectmanager.lua:
```diff
@@ -0,0 +1,301 @@
+end
+
+-- Initialize the MapPinSubject cache for faster reading/writing
+function InitializeMapPinSubjectCache(bForceInitialize:boolean)
```
Hey I went over all the caching logic here and have some thoughts for improvements. Seems like the cache is needed due to serialization/deserialization performance. 

One option is trying other serialization options. Like MetaLua used by "Endgame Map Replay" mod, or DataDumper used by "More Lenses" and "Better Trade Screen".

Another idea is to optimize the usage of the cache. I noticed you are calling init cache for methods like onPinAdded or updatePin, but not for other read methods like isAdjacentToXXX. Thinking about the values being cached, they will only be updated when it is explicitly updated by the current user. I wonder if we can do the following:
1. In GetMapPinSubjectTable, only read from player config if mapPinSubjectCache[playerID] is nil, and the cache is set after the first read from player config. So most reads should read from the cache directly, except the very first read.
2. In UpdateMapPinSubject, take one additional parameter "forceUpdate". All updates get stored to mapPinSubjectCache, and if forceUpdate is true, store to the player config and perform the broadcast. 
3. Add another UpdatePlayerConfig method that forces writing cached data to player config and perform the broadcast.
4. Then in methods like IsValidCityCenterPosition, set forceUpdate to true. But in methods that updates multiple pins at the same time, set forceUpdate to false, and at the end of the method or update, call UpdatePlayerConfig.

The cache is always updated in this case, at least for the current player. I'm not sure how does the current player get play config updates from other players. Will we get it through MapPinManager's Refresh method? If so, before calling BASE_Refresh, you can call a new invalidateCache method to invalid mapPinSubjectCache for all other player ids except the current player's, setting them to nil. So next reads to other players map pin will be reading from player config directly.

Let me know if there's any edge case that I haven't considered.

JamieNyanchi replied:
Thanks for noting the serialization options used by those other mods! I was conflicted on which serialization method to use, and I settled on the one I originally had here before it was all I found when searching on the CivFanatics forum and Civ VI Modding Helpline Discord server.

I tried the ones you shared out and they seem to perform faster than the one I originally chose. I didn't notice any significant difference between MetaLua or DataDumper, so I just chose MetaLua since it seems simpler. As a result, I was able to remove the caching code as it's no longer needed for performance. This reduces the complexity of my added code too.

However, I'm happy to continue the cache discussion if we wish to maybe re-add it for some other reason. I have my responses to your comment below.

---
<details>
<summary>Optimizing the usage of the cache</summary><br>
Is it possible for those other read methods to be called without one of the event handlers being called first (in dmt_yieldcalculator)? I was thinking that it was only necessary to initialize the cache when the events are called because there was no way for the MapPinSubjects to be read/written without those being called first. If it is possible, we could definitely restructure the cache setup to be more like what you proposed. <br><br>
My original idea with the cache was for it to be as minimally invasive as possible. I wasn't sure how you would feel with an always active cache, so I made it so it would only turn on when map pins were about to be updated, and then it would shut down afterwards. If we bring back the cache though and you're okay with it, we could definitely set it up like you proposed.
</details>

---
<details>
<summary>How does the current player get config updates from other players?</summary><br>
To answer this question near the end, the local client gets the player config updates from the other clients through the `Network.BroadcastPlayerInfo()` command. When used, the game (somewhere behind the scenes) updates the player configuration data on every other client and then triggers the `Events.PlayerInfoChanged` event for everyone. In the original `MapPinManager.lua` file, the handler for this event is `OnPlayerInfoChanged()` which then simply calls the `Refresh()` function. So this is how each client gets the data from all other clients, including itself.<br><br>
This is also how I was able to remove the `RefreshMapPinUI()` function. Since the `Events.PlayerInfoChanged` is triggered even when the local player's config is updated, there was no need for that function since the event served the same purpose.
</details>

---
<details>
<summary>Using a new invalidateCache method to invalidate the cache for all other players except the current player</summary><br>
For your idea regarding the new invalidateCache method, if we brought back the cache, we can actually narrow this idea a bit more. The `Events.PlayerInfoChanged` event gives a Player ID when it's triggered, so we could just invalidate the cache for that specific player.<br><br>
My original idea here with the cache was to be minimally invasive, so it would only be active when it needed to be. So the cache is shut off until `Refresh()` was called because that was the primary function that updates all the pins on the UI. I thought that was the best place to put it so the cache wouldn't be active more than it had to be. After that function finished, the cache would shut back down in my code. <br><br>
However, if we expanded it so the cache was active pretty much at all times, then we could just use the event to invalidate the one player and keep the other players' data.
</details>

---

hanweizhang commented on DetailedMapTacks.modinfo:
```diff
@@ -71,7 +68,7 @@
     <!-- Game files -->
     <File>config/dmt_config.xml</File>
-    <File>gameplay/mappinsubjectmanager.lua</File>
+    <File>ui/dmt_mappinsubjectmanager.lua</File>
```
Also wonder how do we consider backward compatibility, i.e. make sure players who has a in-progress game, they won't find the map pin yields missing after updating the mod. 

If we do find serialization performance is a concern, we can always have single player continue to use the existing player property solution, and only apply the new logic if it is a multiplayer game.

JamieNyanchi replied:
Thankfully, backwards compatibility isn't an issue with these changes. Existing games will still have the pin data saved in the player properties, but just won't be used anymore. Instead on the first load for existing saves, it'll check the player configuration for the pin data, find that there's nothing there, and then it'll regenerate all of the data. This is handled in the `OnLocalPlayerTurnBegin` function in `dmt_yieldcalculater.lua`. That same code made to add support for saves made before adding the mod works for this purpose too.

The only concern I could see is that any data saved in the properties from before is just wasted memory now. Though it's pretty insignificant relative to the memory usage of the game itself. Could add a new function to clear it out if it's an issue though.

As for the serialization performance, singleplayer was never an issue. It was only multiplayer that seemed to have a slight regression in performance. It seems to be something behind the scenes, so I can't tell exactly why though. In any case, the new serialization method I implemented based on your suggestion works even better than the previous one. With all that, I don't think it's necessary to apply separate logic for singleplayer vs multiplayer.

JamieNyanchi commented:
Pushed another revision of the PR. I apologize for changing the code fairly significantly each time, but your feedback and suggestions help me make what I feel is a better solution, so I think it's worth it.

Key highlights:
- Changed serialization method from Pickle to MetaLua
- Removed cache system (No longer needed with the increased performance of the new serialization method)
- Moved serialization code to separate file for better code structure

I also responded to your comments you left since last time. I look forward to your new feedback! I think you've helped improve this code considerably, so thank you for that!

hanweizhang commented:
This version looks good to me. Thanks a lot for your patience addressing my comments and your time spent testing the performance. One last question (I promise lol), when you were testing the performance between MetaLua and DataDumper, did you try with DataDumper's "fastMode" flag on? From http://lua-users.org/wiki/DataDumper:

>  The drawback is that it is not the fastest one, mainly because it returns a string and does not write to a temporarily file. Still, there is a "fast" mode with fewer features, when execution time (or memory) is important.

JamieNyanchi commented:
It was no problem at all! I liked our back and forth and I think it resulted in a better solution here.

I honestly can't remember for sure if I tested it with the "fastMode" flag on. I decided to go back and test both MetaLua and DataDumper (with the flag on) just to be sure. I also decided to try testing them with the cache from before just out of curiosity.

Here's what I found:

"One Cluster" means I updated a single district pin surrounded by six other district pins.
"All Pins" means every pin on the map was updated. This occurs at the beginning of every turn. For my testing, I had 35 total pins.
The time given is the range of time elapsed from the beginning of the "UpdatePins" function to the end of it. There are more points in the code where the pin subjects are read, but I used this just as a benchmark since this covers the bulk of the read/write calls.

| Method | One Cluster | One Cluster (Cache) | All Pins | All Pins (Cache) |
| :---: | :---: | :---: | :---: | :---: |
| DataDumper (Singleplayer) | 8-14ms | 1-2ms | 31-41ms | 2-4ms |
| MetaLua (Singleplayer) | 10-15ms | 1-3ms | 35-38ms | 3-4ms |
| DataDumper (Multiplayer) | 14-17ms | 2-3ms | 45-65ms | 4-7ms |
| MetaLua  (Multiplayer) | 12-16ms | 2-5ms | 44-58ms | 5-8ms|

Comparing the two methods, they seem to perform roughly the same, with maybe DataDumper being slightly better on average.

When I removed the cache originally though, I seem to have neglected the performance difference when all the pins on the map are updated at the beginning of each turn. Without a cache, it's a lot of serialize/deserialize calls overall. While each individual call isn't that significant, it can add up. It also gets worse with more and more pins.

The cluster performance should be capped at around the values listed since there can only ever be six adjacent pins. Though it'd probably be worse if the pin you're updating is a City Center pin. I didn't think to test that at the time, but those update every pin within 3 tiles.

What are your thoughts? It's pretty clear that a cache would perform better in all cases, though it would add extra complexity with potentially little benefit if the player doesn't have a lot of pins. For what it's worth, I didn't actually notice these performance impacts while playing. I only notice them when looking at the measured times here.

hanweizhang commented:
Makes sense. I think it'll still be nice to have a cache, an always active cache or only cache at the beginning of every turn, since I've seen people placing hundreds of pins lol. Not necessary though. I'll merge this commit first. Thanks again for your effort!

JamieNyanchi commented:
No problem at all! If you'd like, I could open another PR and we can work on a cache system there. I have some ideas on how to improve the one I originally had here too.
